fast_llm/data/dataset/gpt/sampled.py

```diff
@@ -92,32 +99,93 @@ def __init__(
         self._device = torch.device("cuda" if self._config.gpu else "cpu")
 
         if sampling.cache_directory is None:
-            self._document_shuffling = MemmapArray()
-            self._token_cumsum_shuffled = MemmapArray()
-            self._token_cumsum_unshuffled = MemmapArray()
-            self._yaml_path = None
-            log_main_rank(
-                " > No dataset cache directory provided, building the index map on all ranks."
-                "This may be very inefficient...",
-                log_fn=logger.warning,
-            )
-            self._sample()
+            self._setup_cache()
         else:
             base_path = (
                 sampling.cache_directory / f"{self.name}_ns_{self._num_samples}_sl_{self._sequence_length}"
                 f"_s_{self._config.seed}"
             )
-            # TODO: Names are confusing
-            self._document_shuffling = MemmapArray(base_path.with_name(base_path.name + "_shuffling.npy"))
-            self._token_cumsum_shuffled = MemmapArray(base_path.with_name(base_path.name + "_shuffled_cumsum.npy"))
-            self._token_cumsum_unshuffled = MemmapArray(base_path.with_name(base_path.name + "_unshuffled_cumsum.npy"))
-            self._yaml_path = base_path.with_suffix(".yaml")
+            self._load_cache(base_path)
             # Sample or validate the dataset of a given rank.
             if sampling.distributed.config.rank == sampling.get_next_rank():
                 self._sample()
             # No barrier yet to allow running in parallel.
             # There needs to be one before calling `__getitem__`, normally handled through `GPTData`.
 
+    @abstractmethod
+    def _setup_cache(self):
```

Comment:
Please add return type hints, they are especially important for abstract methods